src/notation/view/notationviewinputcontroller.cpp

```diff
@@ -246,6 +252,85 @@ void NotationViewInputController::setViewMode(const ViewMode& viewMode)
     notation->setViewMode(viewMode);
 }
 
+constexpr qreal scrollStep = .8;
+constexpr qreal thinPadding = 10.0;
+constexpr qreal thickPadding = 25.0;
+
+void NotationViewInputController::moveScreen(int direction)
+{
+    auto notation = currentNotation();
+    if (!notation) {
+        return;
+    }
+    auto scale = m_view->currentScaling();
+    if (notation->viewMode() == ViewMode::LINE) {
+        m_view->moveCanvasHorizontal(m_view->width() * direction * scrollStep / scale);
+    } else {
+        m_view->moveCanvasVertical(m_view->height() * direction * scrollStep / scale);
+    }
+}
+
+void NotationViewInputController::movePage(int direction)
+{
+    auto notation = currentNotation();
+    if (!notation) {
+        return;
+    }
+    if (notation->viewMode() != ViewMode::PAGE) {
+        moveScreen(direction);
+        return;
+    }
+    Page* page = notation->elements()->msScore()->pages().back();
+    if (configuration()->canvasOrientation().val == mu::framework::Orientation::Vertical) {
+        qreal offset = std::min((page->height() + thickPadding) * direction, m_view->toLogical(QPoint()).y() + thickPadding);
+        m_view->moveCanvasVertical(offset);
+    } else {
+        qreal offset = std::min((page->width() + thickPadding) * direction, m_view->toLogical(QPoint()).x() + thickPadding);
+        m_view->moveCanvasHorizontal(offset);
+    }
+}
+
+void NotationViewInputController::nextScreen()
+{
+    moveScreen(-1);
+}
+
+void NotationViewInputController::previousScreen()
+{
+    moveScreen(1);
+}
+
+void NotationViewInputController::nextPage()
+{
+    movePage(-1);
+}
+
+void NotationViewInputController::previousPage()
+{
+    movePage(1);
+}
+
+void NotationViewInputController::startOfScore()
+{
+    auto offset = m_view->toLogical(QPoint());
+    m_view->moveCanvas(offset.x() + thickPadding, offset.y() + thickPadding);
+}
+
+void NotationViewInputController::endOfScore()
+{
+    auto notation = currentNotation();
+    if (!notation) {
+        return;
+    }
+    Ms::MeasureBase* lastMeasure = notation->elements()->msScore()->lastMeasureMM();
+    auto lmRect = lastMeasure->canvasBoundingRect();
+    auto scale = m_view->currentScaling();
+    qreal desiredX = std::max(-thickPadding, lmRect.right() + thickPadding - m_view->width() / scale);
+    qreal desiredY = std::max(-thickPadding, lmRect.bottom() + thickPadding - m_view->height() / scale);
+    auto offset = m_view->toLogical(QPoint());
+    m_view->moveCanvas(offset.x() - desiredX, offset.y() - desiredY);
```

Comment:
Again, much simpler than the MU3 logic and doesn't behave precisely the same way (MU3 seems to scroll unnecessarily far, leaving lots of negative space to the right of the score), but in all cases I tested it ensured the final measure in the last staff (i.e. the bottom right corner of the final measure) was "just" scrolled into view with minimal negative space.